flink-ml-lib/src/main/java/org/apache/flink/ml/classification/ftrl/Ftrl.java

```diff
@@ -0,0 +1,395 @@
+                        .get(0);
+
+        Table onlineModelDataTable = tEnv.fromDataStream(onlineModelData);
+        FtrlModel model = new FtrlModel().setModelData(onlineModelDataTable);
+        ReadWriteUtils.updateExistingParams(model, paramMap);
+        return model;
+    }
+
+    /** Iteration body of ftrl. */
+    public static class FtrlIterationBody implements IterationBody {
+        private final int batchSize;
+        private final double alpha;
+        private final double beta;
+        private final double l1;
+        private final double l2;
+
+        public FtrlIterationBody(int batchSize, double alpha, double beta, double l1, double l2) {
+            this.batchSize = batchSize;
+            this.alpha = alpha;
+            this.beta = beta;
+            this.l1 = l1;
+            this.l2 = l2;
+        }
+
+        @Override
+        public IterationBodyResult process(
+                DataStreamList variableStreams, DataStreamList dataStreams) {
+            DataStream<DenseVector[]> modelData = variableStreams.get(0);
+
+            DataStream<Tuple2<Vector, Double>> points = dataStreams.get(0);
+
+            int parallelism = points.getParallelism();
+            DataStream<DenseVector[]> newModelData =
+                    points.countWindowAll(batchSize)
+                            .apply(new GlobalBatchCreator())
+                            .flatMap(new GlobalBatchSplitter(parallelism))
+                            .rebalance()
+                            .connect(modelData.broadcast())
+                            .transform(
+                                    "ModelDataLocalUpdater",
+                                    TypeInformation.of(DenseVector[].class),
+                                    new FtrlLocalUpdater(alpha, beta, l1, l2))
+                            .setParallelism(parallelism)
+                            .countWindowAll(parallelism)
+                            .reduce(new FtrlGlobalReducer())
+                            .map(
+                                    (MapFunction<DenseVector[], DenseVector[]>)
+                                            value -> new DenseVector[] {value[0]})
+                            .setParallelism(1);
+
+            return new IterationBodyResult(
+                    DataStreamList.of(newModelData),
+                    DataStreamList.of(
+                            modelData
+                                    .map(
+                                            new MapFunction<DenseVector[], LinearModelData>() {
+                                                long iter = 0L;
+
+                                                @Override
+                                                public LinearModelData map(DenseVector[] value) {
+                                                    return new LinearModelData(value[0], iter++);
+                                                }
+                                            })
```

Comment:
Shall we move logics like this to a separated class or method? That could make the code look prettier.